### PR TITLE
kvserver: rm transient fields from CheckConsistencyRequest

### PR DESCRIPTION
**Pull request by pav-kv**
A few fields in CheckConsistencyRequest are used solely to support the
"recursive" CheckConsistency flow which sends another request if the first one
did not compute the diff. This is an implementation detail that should not leak
into the proto API. Having it in the proto increases the possibility of bugs,
as it's hard to reason about this API, and the caller has too much control and
possibility to misconfigure.

Release justification: small fix
Release note: None

### CODE REVIEW

**Review by erikgrinaker — COMMENTED** [review]
Is this change fully backwards compatible with 22.1 clients, or will we need to keep the fields around until 23.1?

**Comment by pavelkalinnikov** [comment]
> Is this change fully backwards compatible with 22.1 clients, or will we need to keep the fields around until 23.1?

This proto is emitted in 2 places:
- Consistency queue. This is local to the process, so proto compatibility is not needed.
- `pkg/sql/consistencychecker/consistency_checker.go`. This can leave the bounds of the process, right? Previously it was sending `WithDiff: true` for full checks, and `WithDiff: false` for the stats checks. After #87375 it doesn't send it (behaviour is equivalent to `WithDiff: false` all the time).

22.1 can briefly send the old `WithDiff: false/true` field to the new version, and will receive response as if it asked for `WithDiff: false` unconditionally. That's fine because this flag only impacts the contents of the `Detail` field, which is for humans running the `crdb_internal.check_consistency()` command.

One wrinkle: will the new binary that has `reserved` tag instead of the actual `WithDiff` field happily take and ignore an unknown field?

**Comment by erikgrinaker** [comment]
Seems fine. Not sure if we actually want to remove these fields though, will have a look.

>One wrinkle: will the new binary that has reserved tag instead of the actual WithDiff field happily take and ignore an unknown field?

Yes.

**Comment by pavelkalinnikov** [comment]
> Seems fine. Not sure if we actually want to remove these fields though, will have a look.

Alternatively, could "soft delete" them, e.g. by saying in comments that they are ignored by the outer API, or marking them as deprecated.

**Review by tbg — APPROVED** [review]
:lgtm: thanks for cleaning this up.

We don't have to worry about `pkg/sql/consistencychecker/consistency_checker.go`, since the `WithDiff` there wasn't useful anyway and is now ignored (with this PR) but also got removed.

Outside of that, I don't think `CheckConsistencyRequest` leaves the process boundary, because here we send it to `repl` which is the local replica, without a `DistSender` routing stack:

https://github.com/cockroachdb/cockroach/blob/9502f33669feaa9b38704eafa36e580da8a97a75/pkg/kv/kvserver/consistency_queue.go#L189-L199

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87389#-:-NBHmK2Z4WqYT6sdG14r:b-77cr7u)* status: :shipit: complete! 1 of 0 LGTMs obtained

**Review by tbg — APPROVED** [review]
> Reviewed 4 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87389)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @pavelkalinnikov)

File: pkg/kv/kvserver/replica_consistency.go [hunk 1]
```diff
@@ -286,9 +284,9 @@ func (r *Replica) CheckConsistency(
 	}
 
 	// No diff was printed, so we want to re-run with diff.
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/replica_consistency.go:286]: s/diff/snapshot/

**Review by erikgrinaker — APPROVED** [review]
This seems like a reasonable cleanup, and appears to be backwards compatible.

Might be worth setting up a mixed-version cluster and run a few consistency checks, just to be sure. I.e., create a local 22.1 cluster using `roachprod`, then stop a few nodes, and start them again using the `master` binary. Happy to help out with specifics if needed.

**Review by pav-kv — COMMENTED** [review]
> Might be worth setting up a mixed-version cluster and run a few consistency checks

Done. The new binary accepts protos with `WithDiff: false/true` from old nodes, and the entire request comes back to the SQL console as if `WithDiff == false`. WAI.

**Comment by pavelkalinnikov** [comment]
bors r=tbg,erikgrinaker

**Comment by pavelkalinnikov** [comment]
bors retry

**Comment by pavelkalinnikov** [comment]
Fixed the test too, it used the deleted fields.

**Comment by pavelkalinnikov** [comment]
bors retry

**Pull request merged**